api/src/main/java/net/kyori/adventure/bossbar/BossBar.java

```diff
@@ -0,0 +1,200 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package net.kyori.adventure.bossbar;
+
+import java.util.Set;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.util.NameMap;
+import org.checkerframework.checker.nullness.qual.NonNull;
+
+/**
+ * A bossbar.
+ */
+public interface BossBar {
+  static @NonNull BossBar of(final @NonNull Component name, final float percent, final @NonNull Color color, final @NonNull Overlay overlay) {
+    return new BossBarImpl(name, percent, color, overlay);
+  }
+
+  /**
+   * Gets the name.
+   *
+   * @return the name
+   */
+  @NonNull Component name();
+
+  /**
+   * Sets the name.
+   *
+   * @param name the name
+   * @return the bossbar
+   */
+  @NonNull BossBar name(final @NonNull Component name);
+
+  /**
+   * Gets the percent.
+   *
+   * <p>The percent is a value between 0 and 1.</p>
+   *
+   * @return the percent
+   */
+  float percent();
+
+  /**
+   * Sets the percent.
+   *
+   * <p>The percent is a value between 0 and 1.</p>
+   *
+   * @param percent the percent
+   * @return the bossbar
+   */
+  @NonNull BossBar percent(final float percent);
+
+  /**
+   * Gets the color.
+   *
+   * @return the color
+   */
+  @NonNull Color color();
+
+  /**
+   * Sets the color.
+   *
+   * @param color the color
+   * @return the bossbar
+   */
+  @NonNull BossBar color(final @NonNull Color color);
+
+  /**
+   * Gets the overlay.
+   *
+   * @return the overlay
+   */
+  @NonNull Overlay overlay();
+
+  /**
+   * Sets the overlay.
+   *
+   * @param overlay the overlay
+   * @return the bossbar
+   */
+  @NonNull BossBar overlay(final @NonNull Overlay overlay);
+
+  /**
+   * Gets the flags.
+   *
+   * @return the flags
+   */
+  @NonNull Set<Flag> flags();
+
+  /**
+   * Sets the flags.
+   *
+   * @param flags the flags
+   * @return the bossbar
+   */
+  @NonNull BossBar flags(final @NonNull Set<Flag> flags);
+
+  /**
+   * Sets the flags.
+   *
+   * @param flags the flags
+   * @return the bossbar
+   */
+  @NonNull BossBar addFlags(final @NonNull Flag@NonNull... flags);
+
+  /**
+   * Sets the flags.
+   *
+   * @param flags the flags
+   * @return the bossbar
+   */
+  @NonNull BossBar removeFlags(final @NonNull Flag@NonNull... flags);
+
+  @NonNull BossBar addListener(final @NonNull Listener listener);
```

Comment:
Now that I think about this again, is a `Listener` interface needed? You could define a `BiConsumer<BossBar, Change>`?

Reply:
I want the method to be named `bossBarChanged` so that it makes more sense if someone decides to implement `Listener` in their own class.

Reply:
https://github.com/KyoriPowered/text/pull/62/commits/fd7ce1a567b36ac5cda35207c90354e398777f5d